cache/in-memory/src/permission.rs

```diff
@@ -0,0 +1,748 @@
+//! ```
+//!
+//! [`ResourceType`]: crate::ResourceType
+
+use super::InMemoryCache;
+use std::{
+    error::Error,
+    fmt::{Display, Formatter, Result as FmtResult},
+};
+use twilight_model::{
+    channel::GuildChannel,
+    guild::Permissions,
+    id::{ChannelId, GuildId, RoleId, UserId},
+};
+use twilight_util::permission_calculator::PermissionCalculator;
+
+/// Error calculating permissions with the information in a cache.
+#[derive(Debug)]
+pub struct ChannelError {
+    kind: ChannelErrorType,
+    source: Option<Box<dyn Error + Send + Sync>>,
+}
+
+impl ChannelError {
+    /// Immutable reference to the type of error that occurred.
+    #[must_use = "retrieving the type has no effect if left unused"]
+    pub const fn kind(&self) -> &ChannelErrorType {
+        &self.kind
+    }
+
+    /// Consume the error, returning the source error if there is any.
+    #[must_use = "consuming the error and retrieving the source has no effect if left unused"]
+    pub fn into_source(self) -> Option<Box<dyn Error + Send + Sync>> {
+        self.source
+    }
+
+    /// Consume the error, returning the owned error type and the source error.
+    #[must_use = "consuming the error into its parts has no effect if left unused"]
+    pub fn into_parts(self) -> (ChannelErrorType, Option<Box<dyn Error + Send + Sync>>) {
+        (self.kind, self.source)
+    }
+
+    /// Create a root error from an error while retrieving a member's roles.
+    fn from_member_roles(member_roles_error: MemberRolesErrorType) -> Self {
+        Self {
+            kind: match member_roles_error {
+                MemberRolesErrorType::MemberMissing { guild_id, user_id } => {
+                    ChannelErrorType::MemberUnavailable { guild_id, user_id }
+                }
+                MemberRolesErrorType::RoleMissing { role_id } => {
+                    ChannelErrorType::RoleUnavailable { role_id }
+                }
+            },
+            source: None,
+        }
+    }
+}
+
+impl Display for ChannelError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self.kind {
+            ChannelErrorType::ChannelUnavailable { channel_id } => f.write_fmt(format_args!(
+                "channel {} is either not in the cache or is not a guild channel",
+                channel_id
+            )),
+            ChannelErrorType::MemberUnavailable { guild_id, user_id } => f.write_fmt(format_args!(
+                "member (guild: {}; user: {}) is not present in the cache",
+                guild_id, user_id
+            )),
+            ChannelErrorType::RoleUnavailable { role_id } => f.write_fmt(format_args!(
+                "member has role {} but it is not present in the cache",
+                role_id
+            )),
+        }
+    }
+}
+
+impl Error for ChannelError {}
+
+/// Type of [`ChannelError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ChannelErrorType {
```

Comment:
These comments differ from those in `RootErrorType`, should they not be the same?